### PR TITLE
[HB-7023] Fix memory leaks occurring when fullscreen ads are shown from an Activity

### PR DESCRIPTION
**Pull request by chauduyphanvu**

### CODE REVIEW

File: AmazonPublisherServicesAdapter/src/main/java/com/chartboost/mediation/amazonpublisherservicesadapter/AmazonPublisherServicesAdapter.kt [hunk 1]
```diff
@@ -864,3 +775,133 @@ class AmazonPublisherServicesAdapter : PartnerAdapter {
         }
     }
 }
+
+private class AdListener(
+    private val continuationRef: WeakReference<CancellableContinuation<Result<PartnerAd>>>,
+    private val request: PartnerAdLoadRequest,
+    private val listenerRef: WeakReference<PartnerAdListener?>,
+) : DTBAdInterstitialListener {
+    private lateinit var fullscreenAd: DTBAdInterstitial
```
Comment (bichenwang) [inline AmazonPublisherServicesAdapter/src/main/java/com/chartboost/mediation/amazonpublisherservicesadapter/AmazonPublisherServicesAdapter.kt:784]: this holds onto the activity, though. which should be fine, maybe?
what if you load an amazon placement, rotate the screen to kill the current activity, and try to show that ad?
also, do you need this to be lateinit? Why not just make it `DTBAdInterstitial?`

Comment (chauduyphanvu) [reply AmazonPublisherServicesAdapter/src/main/java/com/chartboost/mediation/amazonpublisherservicesadapter/AmazonPublisherServicesAdapter.kt]: kk changed to the nullable type. I don't have fills anymore so can't test the rotation thing.

File: AmazonPublisherServicesAdapter/src/main/java/com/chartboost/mediation/amazonpublisherservicesadapter/AmazonPublisherServicesAdapter.kt [hunk 2]
```diff
@@ -823,15 +732,17 @@ class AmazonPublisherServicesAdapter : PartnerAdapter {
...
-                        if (continuation.isActive) {
-                            continuation.resume(result)
-                        }
-                    }
+                    val continuationWeakRef = WeakReference(continuation)
 
                     onShowSuccess = {
                         PartnerLogController.log(SHOW_SUCCEEDED)
-                        resumeOnce(Result.success(partnerAd))
+                        continuationWeakRef.get()?.let {
```
Comment (bichenwang) [inline AmazonPublisherServicesAdapter/src/main/java/com/chartboost/mediation/amazonpublisherservicesadapter/AmazonPublisherServicesAdapter.kt:739]: oh then again there are only weak references here, so it should be fine

Comment (chauduyphanvu) [reply AmazonPublisherServicesAdapter/src/main/java/com/chartboost/mediation/amazonpublisherservicesadapter/AmazonPublisherServicesAdapter.kt:739]: yep, nothing here so should be good

File: AmazonPublisherServicesAdapter/src/main/java/com/chartboost/mediation/amazonpublisherservicesadapter/AmazonPublisherServicesAdapter.kt [hunk 3]
```diff
@@ -864,3 +775,140 @@ class AmazonPublisherServicesAdapter : PartnerAdapter {
...
+                    "Unable to fire onPartnerAdDismissed for Amazon Publisher Services adapter. Listener is null.",
+                )
+            }
+        }
+    }
+
+    override fun onImpressionFired(adView: View?) {
+        CoroutineScope(Main).launch {
+            PartnerLogController.log(DID_TRACK_IMPRESSION)
+            onShowSuccess()
```
Comment (bichenwang) [inline AmazonPublisherServicesAdapter/src/main/java/com/chartboost/mediation/amazonpublisherservicesadapter/AmazonPublisherServicesAdapter.kt:878]: oh as safety, should we clear the lambda after calling this?

Comment (chauduyphanvu) [reply AmazonPublisherServicesAdapter/src/main/java/com/chartboost/mediation/amazonpublisherservicesadapter/AmazonPublisherServicesAdapter.kt:878]: I tested that that but saw no differences with anything 🤔

**Review by bichenwang — APPROVED** [review]

**Review by cb-jpadilla — APPROVED** [review]

**Pull request merged**